### PR TITLE
Productions

### PR DESCRIPTION
**Pull request by mrzelak**

### CODE REVIEW

File: src/basic_graph.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    nodes = [node_1, node_2, node_3, node_4]
+
+    graph.add_node_collection(nodes)
+    graph.add_node(node_q)
+
+    # add `edges' of hyperedge
+    for node in nodes:
+        attr = EdgeAttrs(kind='q', flag=hanging)
+        edge = Edge(node.handle, node_q.handle, attr)
+        graph.add_edge(edge)
```
Comment (kkafar) [inline src/basic_graph.py:37]: Please use `add_q_hyperedge` here. This code does not ensure single edge handle for all edges that belong to this hyperedge.

File: src/basic_graph.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        node_4 = Node(NodeAttrs('v', 0, 1, False))
+        node_5 = Node(NodeAttrs('v', 1.3, 0.5, False))
+        node_p = Node(NodeAttrs('p', 0.5, 0.5, select_central))
+
+    nodes = [node_1, node_2, node_3, node_4, node_5]
+    graph.add_node_collection(nodes)
+    attr = EdgeAttrs('p', flag=select_central)
+    graph.add_p_hyperedge(nodes, attr, node_p.handle, (0.5, 0.5)) 
+
+    return graph, nodes
```
Comment (kkafar) [inline src/basic_graph.py:74]: Is there a reason to return tuple here, and not do it in other methods?

File: src/production/production_10.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        self.rev_mapping: Dict[NodeHandle, NodeHandle] | None = None
+
+    def get_lhs(self) -> Graph:
+        return self.lhs
+
+    def reset(self):
+        self.rev_mapping = None
+
+    def requires_monomorphism(self):
+        return True
```
Comment (kkafar) [inline src/production/production_10.py:22]: <3

File: src/production/production_16.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        ]
+        # get inner node, the hyperedge
+        node_p = graph.node_for_handle(rev_mapping[5])
+        x = node_p.attrs.x
+        y = node_p.attrs.y
+
+        graph.remove_p_hyperedge(node_p.handle)
+
+        attr = EdgeAttrs('p', flag=True)
+        graph.add_p_hyperedge(vertices, attr, node_p.handle, (x, y))
```
Comment (kkafar) [inline src/production/production_16.py:50]: I didn't reason much what's going on here, but if you only removing the p edge, changing flag value & reinserting it into the same place: you can just edit `node_p.attrs.flag = True` without adding/removing the edge and it should work as you get reference to original attributes.

If it is not the case, just ignore this comment.

**Review by kkafar — COMMENTED** [review]
Just one remark, great job overall!

File: src/basic_graph.py (hunk 1 above)
Comment (mrzelak) [reply src/basic_graph.py:37]: valid point

File: src/basic_graph.py (hunk 2 above)
Comment (mrzelak) [reply src/basic_graph.py:74]: _nodes_ are needed in _basic_pentagon_ to shuffle node order to match order presented in the spec, while in the spec node order of star (used in P16) has order of nodes present in _basic_star5_.

File: src/production/production_16.py (hunk 4 above)
Comment (mrzelak) [reply src/production/production_16.py:50]: This is exactly the case, and I'm almost sure I tried this approach and failed for some unknown reason so it's why this spaghetti code is present. Nevertheless, now it's perfectly working and refactoring code in this way

**Review by kkafar — APPROVED** [review]
Great job!

**Pull request merged**